tests/integration/vxlan_test.py

```diff
@@ -0,0 +1,110 @@
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 2.1 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+#
+
+import time
+
+import pytest
+
+import libnmstate
+
+from libnmstate.error import NmstateVerificationError
+from libnmstate.schema import Interface
+
+from .testlib import assertlib
+from .testlib.bondlib import bond_interface
+from .testlib.vxlan import VxlanState
+from .testlib.vxlan import vxlan_interfaces
+from .testlib.vxlan import vxlans_up
+from .testlib.vxlan import vxlans_down
+
+VXLAN1_ID = 101
+VXLAN2_ID = 102
+
+
+def test_add_and_remove_vxlan(eth1_up):
+    ifname = eth1_up[Interface.KEY][0][Interface.NAME]
+    with vxlan_interfaces(
+        VxlanState(id=VXLAN1_ID, base_if=ifname, remote='192.168.100.1')
+    ) as desired_state:
+        assertlib.assert_state(desired_state)
+
+    vxlan1_ifname = desired_state[Interface.KEY][0][Interface.NAME]
+    assertlib.assert_absent(vxlan1_ifname)
+
+
+def test_add_and_remove_two_vxlans_on_same_iface(eth1_up):
+    ifname = eth1_up[Interface.KEY][0][Interface.NAME]
+    with vxlan_interfaces(
+        VxlanState(id=VXLAN1_ID, base_if=ifname, remote='192.168.100.1'),
+        VxlanState(id=VXLAN2_ID, base_if=ifname, remote='192.168.100.2'),
+    ) as desired_state:
+        assertlib.assert_state(desired_state)
+
+    vxlan_interfaces_name = (
+        i[Interface.NAME] for i in desired_state[Interface.KEY]
+    )
+    assertlib.assert_absent(vxlan_interfaces_name)
+
+
+def test_rollback_for_vxlans(eth1_up):
+    ifname = eth1_up[Interface.KEY][0][Interface.NAME]
+    current_state = libnmstate.show()
+    desired_state = vxlans_up(
+        [
+            VxlanState(id=VXLAN1_ID, base_if=ifname, remote='192.168.100.1'),
+            VxlanState(id=VXLAN2_ID, base_if=ifname, remote='192.168.100.2'),
+        ]
+    )
+    desired_state[Interface.KEY][1]['invalid_key'] = 'foo'
+    with pytest.raises(NmstateVerificationError):
+        libnmstate.apply(desired_state)
+
+    time.sleep(5)  # Give some time for NetworkManager to rollback
+    current_state_after_apply = libnmstate.show()
+    assert current_state == current_state_after_apply
+
+
+def test_set_vxlan_iface_down(eth1_up):
+    ifname = eth1_up[Interface.KEY][0][Interface.NAME]
+    vxlan = VxlanState(id=VXLAN1_ID, base_if=ifname, remote='192.168.100.1')
+    with vxlan_interfaces(vxlan):
+        libnmstate.apply(vxlans_down([vxlan]))
+
+    assertlib.assert_absent(vxlan.name)
```

Comment:
I think you want to check that the ‘down’ was applied (i.e. removed). If so, this assert needs to be indented into the ‘with’ block.

Reply:
absent is good here to check that vxlan no longer exists, we can also add a down check within with.

Reply:
The ‘down’ should remove it.
There is no need to assert what was already tested elsewhere, although I do not mind too much if you do.

The test says it checks setting the iface down, so it makes sense to assert it IMO. How will you know if it worked?

Reply:
ack, asserting within context manager.